PHP 7.4.9 Revision 2 and Composer Update

timothygwebb opened the pull request:
All Changes are documented in code.
Similar changes single comment grouped by line view in window.

rndevuk commented:
hey tim. glad to see progress being made. these merge conflicts need to be fixed, also the bb-vendor directory should be removed before pr's/commits.

rndevuk commented:
actually remove the bb-vendor dir actually fixes the majority of the merge conflicts.

timothygwebb commented:
Those fixes in the bb-vendor are required for the php upgrade. It is fixing
the arrays that are in {} to (). Due to the following unless we as a group
decide to start coding using json reference as follows. That's a JSON
<http://www.json.org/> string representation of an array, use json_decode():

$array = json_decode($string);
print_r($array);

Yields:

Array(
    [0] => Array
        (
            [0] => Array
                (
                    [0] => Name
                    [1] => ID
                )
            [1] => Array
                (
                    [0] => 12
                )
        )
    [1] => Array
        (
            [0] => Test
            [1] => 78
        )
    [2] => 0)

If it had any { } that would be an object and decoded as a stdClass object
in PHP unless you pass true to json_decode() to force an array.

Since it's structured as a PHP array (as of PHP 5.4), this works as well
(don't use eval on untrusted data). There is absolutely no reason to do
this, it's just for fun:

eval("\$array = $string;");
print_r($array);


On Sun, Sep 6, 2020, 7:40 AM Ryan Nolan <notifications@github.com> wrote:

> actually remove the bb-vendor dir actually fixes the majority of the merge
> conflicts.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/boxbilling/boxbilling/pull/634#issuecomment-687765628>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD24K7MUXCA2JDP5NCXMIM3SENYJHANCNFSM4Q2LDKYQ>
> .
>

timothygwebb commented:
Array and string offset access using curly braces ¶
<https://www.php.net/manual/en/migration74.deprecated.php#migration74.deprecated.core.array-string-access-curly-brace>

The array and string offset access syntax using curly braces is deprecated.
Use $var[$idx] instead of $var{$idx}.

On Sun, Sep 6, 2020, 9:19 AM Mr. Timothy G Webb Sr. <
notifications@github.com> wrote:

> Those fixes in the bb-vendor are required for the php upgrade. It is fixing
> the arrays that are in {} to (). Due to the following unless we as a group
> decide to start coding using json reference as follows. That's a JSON
> <http://www.json.org/> string representation of an array, use
> json_decode():
>
> $array = json_decode($string);
> print_r($array);
>
> Yields:
>
> Array(
> [0] => Array
> (
> [0] => Array
> (
> [0] => Name
> [1] => ID
> )
> [1] => Array
> (
> [0] => 12
> )
> )
> [1] => Array
> (
> [0] => Test
> [1] => 78
> )
> [2] => 0)
>
> If it had any { } that would be an object and decoded as a stdClass object
> in PHP unless you pass true to json_decode() to force an array.
>
> Since it's structured as a PHP array (as of PHP 5.4), this works as well
> (don't use eval on untrusted data). There is absolutely no reason to do
> this, it's just for fun:
>
> eval("\$array = $string;");
> print_r($array);
>
>
> On Sun, Sep 6, 2020, 7:40 AM Ryan Nolan <notifications@github.com> wrote:
>
> > actually remove the bb-vendor dir actually fixes the majority of the
> merge
> > conflicts.
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/boxbilling/boxbilling/pull/634#issuecomment-687765628>,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AD24K7MUXCA2JDP5NCXMIM3SENYJHANCNFSM4Q2LDKYQ
> >
> > .
> >
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/boxbilling/boxbilling/pull/634#issuecomment-687787878>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD24K7PAFB5B5OARZYMLLEDSEOD7PANCNFSM4Q2LDKYQ>
> .
>

rndevuk commented:
I understand what you are doing but you can not edit files in the bb-vendor.
1 composer update and it will break someones install.

if there is no other solution then then next best thing would be to fork the packages that need editing pushing the changes to the origin repo and then while waiting for them to be approved set composer to download from the fork you created....

timothygwebb commented:
Ryan,

So just to clarify. I will need to make a pull request to the corresponding
package repository that needs modification and then have boxbilling use the
corresponding release or tag that will not break the install.

On Sun, Sep 6, 2020, 1:40 PM Ryan Nolan <notifications@github.com> wrote:

> I understand what you are doing but you can not edit files in the
> bb-vendor.
> 1 composer update and it will break someones install.
>
> if there is no other solution then then next best thing would be to fork
> the packages that need editing pushing the changes to the origin repo and
> then while waiting for them to be approved set composer to download from
> the fork you created....
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/boxbilling/boxbilling/pull/634#issuecomment-687851718>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD24K7NV3U5WURSTF5YRCQLSEPCPTANCNFSM4Q2LDKYQ>
> .
>

rndevuk commented:
@timothygwebb that is correct. but while you are waiting for the corresponding repo to accept your PR. we can tell composer to download from your fork of that repo.

timothygwebb commented:
I will need help with telling composer to use my repo. I think I know what
needs to be done. But I just want to make sure I do it right. I will talk
more with you later on discord about this and the fine detail of how to do
it.

On Sun, Sep 6, 2020, 3:40 PM Ryan Nolan <notifications@github.com> wrote:

> @timothygwebb <https://github.com/timothygwebb> that is correct. but
> while you are waiting for the corresponding repo to accept your PR. we can
> tell composer to download from your fork of that repo.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/boxbilling/boxbilling/pull/634#issuecomment-687881336>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD24K7OZKCW6WK4E3RVXYCDSEPQS7ANCNFSM4Q2LDKYQ>
> .
>

samapraku commented:
@timothygwebb first of all, note that your efforts are appreciated.
I wanted to review the updates you have done but it is a bit difficult to identify the specific files you updated. I think the previous pull requests that were merged had the large number of files issue but since has already been merged it is going to appear in subsequent pull requests. I must say that reviewing the new updates is a bit of a challenge.

**Point:**
You have added a new composer.json file at src/bb-library/composer.json and looks like you are using that to add a package.
Since it is an external package, shouldn't it be installed and managed together with all the other packages in src/composer.json?

samapraku commented:
I'm sure that the need to modify files in the bb-vendor directory can be avoided if we could all properly follow the updates you have made in order to make inputs. We are all aware that boxbilling's codebase is old so if the solution is to fix the code instead, then we can all work on fixing the code.

timothygwebb commented:
First off I will explain why I want to include the bb-vendor directory.
If you are given this software to download and install to begin with  it does not come with the bb-vendor directory here is what you get.

/httpdocs/boxbilling/src/bb-vendor does not exist and could not be created. 

unless you are a superuser doing an install on the operating system no rights are given to create the directory upon download of zip or clone.

So one has to issue a command as such for the update/install to even occur mkdir ~(home)/httpdocs/boxbilling/src/bb-vendor and then it will allow you to update the packages.

Using the file manager on my server if I was a client I would encounter the following. 

/httpdocs/boxbilling/src/bb-vendor does not exist and could not be created. 

I know the issue is that the application requires root permissions on my server to be able to access anything past the /var/www directory.

So I issue the following  sudo mkdir bb-vendor thinking hey I am pretty smart but maybe not because I encounter many more errors for each composer update that occurs because it cant create the directories needed to do the udpates. So I go through the process of creating all the needed directories to install the composer updates needed for my php version and it finally succeeds. So we either need to include a starting bb-vendor package folder or instruct/force boxbilling to install composer packages and do an update check because just downloading the zip/cloning does not allow for an easy install on a blank system install just running the defaults.

So I finally got smart and did a little research and followed the directions on the following website.

https://getcomposer.org/doc/faqs/how-to-install-composer-programmatically.md

We need to include the script below to be called during sometime in the install given a check on system type and root directory could be anything from httpdocs, public_html, html, and the assoundrious that contains the boxbilling src directyory to instal the .phar file in to execute the statement below.

Then I executed the following sudo wget https://raw.githubusercontent.com/composer/getcomposer.org/fb07bd70d93cf733c1c94aef7fc68502f81aca2b/web/installer -O - -q | php -- --quiet && ./composer.phar install.

It installed all the packages specified in composer.json and composer.lock and I worked on from there updating what I needed until I had a clean update finally ending up with the following.

I did a manual edit on the .lock files and .json file for guzzle and geoip2. 

The rest was done programmatically.

Loading composer repositories with package information
Updating dependencies (including require-dev)
Package operations: 7 installs, 2 updates, 1 removal
  - Removing guzzle/guzzle (v3.9.3)
  - Installing composer/ca-bundle (1.2.8): Downloading (100%)         
  - Installing maxmind/web-service-common (v0.7.0): Downloading (100%)         
  - Updating geoip2/geoip2 (v2.1.0 => v2.10.0): Downloading (100%)         
  - Installing symfony/polyfill-php80 (v1.18.1): Downloading (100%)         
  - Installing psr/event-dispatcher (1.0.0): Downloading (100%)         
  - Installing symfony/event-dispatcher-contracts (v2.1.3): Downloading (100%)         
  - Installing symfony/deprecation-contracts (v2.1.3): Downloading (100%)         
  - Updating symfony/event-dispatcher (v2.8.52 => v5.1.5): Downloading (100%)         
  - Installing guzzlehttp/guzzle (v3.8.1): Downloading (100%)         
Writing lock file
Generating autoload files
35 packages you are using are looking for funding.
Use the `composer fund` command to find out more!


The majority of the issues I encountered is the fact I had to do a sudo to install boxbilling into the correct directory giving all the files root:root permissions so I executed the following chown -R boxbilling webmaster:psacln and finished updating the necessary composer packages that had issues and there are a lot of requirements that are not even included in the currrent installation that are recommended as dependencies of the other packages.

Would you like to include them as well?

maxmind-db/reader suggests installing ext-bcmath (bcmath or gmp is required for decoding larger integers with the pure PHP decoder)
maxmind-db/reader suggests installing ext-gmp (bcmath or gmp is required for decoding larger integers with the pure PHP decoder)
maxmind-db/reader suggests installing ext-maxminddb (A C-based database decoder that provides significantly faster lookups)
symfony/event-dispatcher suggests installing symfony/dependency-injection
symfony/event-dispatcher suggests installing symfony/http-kernel
guzzle/guzzle suggests installing guzzlehttp/guzzle (Guzzle 5 has moved to a new package name. The package you have installed, Guzzle 3, is deprecated.)
phpmailer/phpmailer suggests installing hayageek/oauth2-yahoo (Needed for Yahoo XOAUTH2 authentication)
phpmailer/phpmailer suggests installing league/oauth2-google (Needed for Google XOAUTH2 authentication)
phpmailer/phpmailer suggests installing psr/log (For optional PSR-3 debug logging)
phpmailer/phpmailer suggests installing stevenmaguire/oauth2-microsoft (Needed for Microsoft XOAUTH2 authentication)
twig/extensions suggests installing symfony/translation (Allow the time_diff output to be translated)
sebastian/global-state suggests installing ext-uopz (*)
phpunit/php-code-coverage suggests installing ext-pcov (*)
phpunit/php-code-coverage suggests installing ext-xdebug (*)
phpunit/phpunit suggests installing ext-soap (*)
phpunit/phpunit suggests installing ext-xdebug (*)

![Screenshot from 2020-09-07 00-30-40](https://user-images.githubusercontent.com/16106877/92348591-83365b80-f0a1-11ea-8d80-504c8f6fd41a.png)


![Screenshot from 2020-09-07 00-17-37](https://user-images.githubusercontent.com/16106877/92347985-9ea06700-f09f-11ea-8bb1-01b4bb386327.png)

![Screenshot from 2020-09-07 00-11-47](https://user-images.githubusercontent.com/16106877/92347706-ec689f80-f09e-11ea-8cbd-dcb4811f3ae3.png)

![Screenshot from 2020-09-07 00-02-53](https://user-images.githubusercontent.com/16106877/92347341-cdb5d900-f09d-11ea-8cc8-a9cdf455d7f7.png)

So with all that said to get php7.4.9 and composer updated as of right now I am going to branch off to a composer branch and a php 7.4.9 branch to allow easier updating and the rest of my comments will be done on the other pull request. See attached screenshot for final results of composer updates.
![Screenshot from 2020-09-07 01-17-24](https://user-images.githubusercontent.com/16106877/92350871-415ce380-f0a8-11ea-8cd4-1d7e46a1072f.png)

samapraku commented:
The bb-vendor directory isn't included because it is not part of the codebase. The directory is created based on the composer.json and composer.lock files.  When a version of the application is released, bb-vendor will be generated and bundled with the released setup files so users do do not have to do composer install or composer update.

The installation steps you have described above is specific to your development environment and doesn't apply for all users. You need to composer in order to install the packages in composer.json or composer.lock files so if you don't have composer then you have to install it.

samapraku commented:
re-post:

You have added a new composer.json file at src/bb-library/composer.json and looks like you are using that to add a package.
Since it is an external package, shouldn't it be installed and managed together with all the other packages in src/composer.json?

rndevuk commented:
In regards to composer not being able to be used by some potential users. we can solve that. @fordnox could setup an automated action that does this upon release... which means only the releases would have the bb-vendor.

but bb-vendor should no be in the repo.

timothygwebb commented:
> re-post:
> 
> You have added a new composer.json file at src/bb-library/composer.json and looks like you are using that to add a package.
> Since it is an external package, shouldn't it be installed and managed together with all the other packages in src/composer.json?

@samapraku  Both issues you continue to mention have been fixed in the new pull requests. I have also fixed the original pull requests and removed the requested directory of bb-vendor. Please review the requests and realize that I only show the information regarding bb-vendor because I like to use a product with ease and I would hate to have any customer of mine that actually pays for a product have any install issues such as mentioned in the prior pull request. I am sure @fordnox  would agree with this as well since I pay for this product.

samapraku commented on src/bb-library/Box/Log.php:
```diff
@@ -41,8 +41,13 @@ class Box_Log
          self::INFO     => 'INFO',
          self::DEBUG    => 'DEBUG',
     );
+<<<<<<< HEAD
```
Conflict resolution required.
@timothygwebb kindly revert this file to the original state. 
I have found what is causing the "Trying to access array offset on value of type null"  error and the solution is not in this particular file. 
I will submit a PR to fix that.

timothygwebb replied:
@samapraku  Will do and thanks.

fordnox commented:
use smaller pull requests https://about.gitlab.com/handbook/engineering/workflow/iteration/#how-to-split-a-merge-request
